### PR TITLE
Fixed #32899 -- Added note about avoiding non-dict objects in JsonResponse docs.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/32899

### CODE REVIEW

**Comment by felixxm** [comment]
@graingert Do you want to take a look?

**Comment by graingert** [comment]
Hmm it's a tricky concept to explain here. It might need some examples in the main documentation

**Review by graingert — APPROVED** [review]
This is a definite improvement but I think these docs could be clarified further

**Comment by tim-mccurrach** [comment]
I'm very much in two minds about this:
- On the one hand, it's very useful advice. 
- However, this just feels like generally-good-advice about designing an API. I'm not sure why it belongs in the Django docs. There are plenty of places we add useful bits of information, but they are normally far more Django-specific, or to highlight unusual quirks/pitfalls. 

I'm definitely not against including this, but I do wonder if it sets a dangerous precedent.

Either way; the section "Serializing non-dictionary objects" is currently about why you need to pass `safe=False`. This is separate to the issue of forward-compatibility, so any discussion about forward-compatibility should be in its own paragraph so that these two issues are not conflated. 

Maybe it could be added as a best practice admonition?

File: django/http/response.py [hunk 1]
```diff
@@ -588,8 +588,9 @@ class JsonResponse(HttpResponse):
     An HTTP response class that consumes data to be serialized to JSON.
 
     :param data: Data to be dumped into json. By default only ``dict`` objects
-      are allowed to be passed due to a security flaw before ECMAScript 5. See
-      the ``safe`` parameter for more information.
+      are allowed to be passed due to a security flaw before ECMAScript 5.
+      Also, dict objects make the expansion of the response simpler. See the
+      ``safe`` parameter for more information.
```
Comment (cjerdonek) [inline django/http/response.py:593]: IMO, it's not clear to me what the second-last sentence means. And then when you go to the `safe` parameter section for more info, it doesn't expand on that, it just repeats the same sentence.

Is it trying to say that even though only `dict` objects are allowed, it isn't a bad thing?

Comment (hramezani) [reply django/http/response.py:593]: There is a description about how to disable the default behaviour (`By default only ``dict`` objects are allowed to be passed`) in the `safe` parameter section.

> Is it trying to say that even though only dict objects are allowed, it isn't a bad thing?

Yes, I think so

Comment (graingert) [reply django/http/response.py:593]: No, you should only return dict objects in json responses unless you're reimplementing an existing protocol that forces a JSON Array on you

Comment (graingert) [reply django/http/response.py:593]: This was historically because of a security issue, but now it's because arrays do not support forward comparability

**Comment by felixxm** [comment]
> Either way; the section "Serializing non-dictionary objects" is currently about why you need to pass `safe=False`. This is separate to the issue of forward-compatibility, so any discussion about forward-compatibility should be in its own paragraph so that these two issues are not conflated.

I don't agree, this section is rather a "how-to" serialize objects other than `dict`, so it looks reasonable to me to add here a short note about why it may not be the best idea.

File: django/http/response.py (hunk 1 above)
Comment (felixxm) [reply django/http/response.py:593]: After reconsideration, I would leave `JsonResponse`'s docstring and the warning in _"Serializing non-dictionary objects"_ untouched. Instead we can add a small note in the main text, e.g.
```diff
diff --git a/docs/ref/request-response.txt b/docs/ref/request-response.txt
index a34880a4be..4e09e3ec5c 100644
--- a/docs/ref/request-response.txt
+++ b/docs/ref/request-response.txt
@@ -1092,6 +1092,10 @@ parameter to ``False``::
 
 Without passing ``safe=False``, a :exc:`TypeError` will be raised.
 
+Note that an API based on ``dict`` objects is more extensible, flexible, and
+easier to maintain forwards compatibility. Therefore, you should avoid using
+non-dict objects in JSON-encoded response.
+
 .. warning::
 
     Before the `5th edition of ECMAScript
```

**Comment by felixxm** [comment]
@graingert Can you take a look?

**Review by graingert — APPROVED** [review]

**Comment by felixxm** [comment]
Thanks for all reviews :+1:

**Pull request merged**